docs/docs/reference/metaprogramming/inline.md

```diff
@@ -0,0 +1,371 @@
+an `inline val` must itself be a [constant expression](#the-definition-of-constant-expression).
+Used in this way, `inline` is equivalent to Java and Scala 2's `final`. `final` meaning
+"constant" is still supported in Dotty, but will be phased out.
+
+The `Logger` object contains a definition of the **inline method** `log`.
+This method will always be inlined at the point of call.
+
+In the inlined code, an if-then-else with a constant condition will be rewritten
+to its then- or else-part. Consequently, in the `log` method above
+`if (Config.logging)` with `Config.logging == true` will rewritten into its then-part.
+
+Here's an example:
+
+```scala
+def factorial(n: BigInt): BigInt =
+  log(s"factorial($n)") {
+    if (n == 0) 1
+    else n * factorial(n - 1)
+  }
+```
+
+If `Config.logging == false`, this will be rewritten (simplified) to
+
+```scala
+def factorial(n: BigInt): BigInt = {
+  /* parameters of log passed by-name (1) */
+  def msg = s"factorial($n)"
+  def op =
+    if (n == 0) 1
+    else n * factorial(n - 1)
+
+  /* inlined body of log  (2) */
+  op
+}
+```
+
+and if `true` it will be rewritten in the code below:
+
+```scala
+def factorial(n: BigInt): BigInt = {
+  /* parameters of log passed by-name (1) */
+  def msg = s"factorial($n)"
+  def op =
+    if (n == 0) 1
+    else n * factorial(n - 1)
+
+  /* inlined body of log  (2) */
+  println(s"${"  " * indent}start $msg")
+  indent += 1
+  val result = op
+  indent -= 1
+  println(s"${"  " * indent}$msg = $result")
+  result
+}
+```
+
+Note (1) that the arguments corresponding to the parameters `msg` and `op` of
+the inline method `log` are defined before the inlined body (which is in this
+case simply `op` (2)). By-name parameters of the inline method correspond to
+`def` bindings whereas by-value parameters correspond to `val` bindings. So if
+`log` was defined like this:
+
+```scala
+inline def log[T](msg: String)(op: => T): T = ...
+```
+
+we'd get
+
+```scala
+val msg = s"factorial($n)"
+```
+
+instead. This behavior is designed so that calling an inline method is
+semantically the same as calling a normal method: By-value arguments
+are evaluated before the call whereas by-name arguments are evaluated
+each time they are referenced. As a consequence, it is often
+preferable to make arguments of inline methods by-name in order to
+avoid unnecessary evaluations. Additionally, in the code above, our
+goal is to print the result after the evaluation of `op`. Imagine, if we were
+printing the duration of the evaluation between the two prints.
+
+For instance, here is how we can define a zero-overhead `foreach` method
+that translates into a straightforward while loop without any indirection or
+overhead:
+
+```scala
+inline def foreach(op: => Int => Unit): Unit = {
+  var i = from
+  while (i < end) {
+    op(i)
+    i += 1
+  }
+}
+```
+
+By contrast, if `op` is a call-by-value parameter, it would be evaluated separately as a closure.
```

Comment:
The reason why we do not generate the closure is that when we only refer once to a by-name parameter once we can inline the code directly. Which then allows us to beta reduce the call. But if we use it more than one the def will be created to avoid duplication of code. All this is not clear at this point.